crates/publish/src/oci/client.rs

```diff
@@ -248,3 +308,87 @@ impl Client {
         }
     }
 }
+
+#[derive(Serialize, Deserialize)]
+struct AuthConfig {
+    /// Map between registry server and base64 encoded username:password credential set.
+    pub auths: HashMap<String, String>,
+}
+
+impl AuthConfig {
+    /// Load the authentication configuration from the default location
+    /// ($XDG_CONFIG_HOME/fermyon/registry-auth.json).
+    pub async fn load_default() -> Result<Self> {
+        // TODO: add a way to override this path.
+        match Self::load(&Self::config_path()?).await {
+            Ok(s) => Ok(s),
+            Err(_) => Ok(Self {
+                auths: HashMap::new(),
+            }),
+        }
+    }
+
+    /// Save the authentication configuration to the default location
+    /// ($XDG_CONFIG_HOME/fermyon/registry-auth.json).
+    pub async fn save_default(&self) -> Result<()> {
+        self.save(&Self::config_path()?).await
+    }
+
+    /// Insert the new credentials into the auths file, with the server as the key and base64
+    /// encoded username:password as the value.
+    pub fn insert(
+        &mut self,
+        server: impl AsRef<str>,
+        username: impl AsRef<str>,
+        password: impl AsRef<str>,
+    ) -> Result<()> {
+        let encoded = base64::Engine::encode(
+            &base64::engine::general_purpose::STANDARD,
+            format!("{}:{}", username.as_ref(), password.as_ref()),
+        );
+        self.auths.insert(server.as_ref().to_string(), encoded);
+
+        Ok(())
+    }
+
+    fn config_path() -> Result<PathBuf> {
+        Ok(dirs::config_dir()
+            .context("Cannot find configuration directory")?
+            .join("fermyon")
+            .join("registry-auth.json"))
+    }
+
+    /// Get the registry authentication for a given registry from the default location.
+    async fn get_auth_from_default(server: impl AsRef<str>) -> Result<RegistryAuth> {
+        let auths = Self::load_default().await?;
+        let encoded = match auths.auths.get(&server.as_ref().to_string()) {
+            Some(e) => e,
+            None => bail!("Cannot find credentials"),
+        };
+
+        let bytes = base64::Engine::decode(&base64::engine::general_purpose::STANDARD, encoded)?;
+        let decoded = std::str::from_utf8(&bytes)?;
+        let parts: Vec<&str> = decoded.splitn(2, ':').collect();
+
+        tracing::trace!("Decoded registry credentials from the Spin configuration.");
+        Ok(RegistryAuth::Basic(
+            parts
+                .first()
+                .context("expected username as first element of the decoded auth")?
+                .to_string(),
+            parts
+                .get(1)
+                .context("expected secret as second element of the decoded auth")?
+                .to_string(),
+        ))
+    }
+
+    async fn load(p: &Path) -> Result<Self> {
+        let contents = tokio::fs::read_to_string(&p).await?;
```

Comment:
Does this need context?  Currently the only place we consume it discards all errors, but if we ever expect errors to be surfaced (and if not, why not return an Option), then the ol' `file or directory does not exist os error 2` could really do with some context for which file it was.